### PR TITLE
[wip, test CI] Add IterableDataset

### PR DESCRIPTION
**Pull request by ssnl**
1. Add `IterableDataset`.
2. Support non batched loading of traditional map-like dataset. This is useful in doing bulk loading.
3. So we have three data loader mods: `Iterable` (newly added), `Map` (newly added), and `MapWithBatchedRead` (old).

    1. `Iterable` if the `dataset` is an instance of `IterableDataset`
    2. `Map` if `batch_size` is `None`
    3. `MapWithBatchedRead` is chosen otherwise.

3. Refactor `DataLoaderIter` into two classes, `_SingleProcessDataLoaderIter` and `_MultiProcessingDataLoaderIter`. Rename some methods to be more generic, e.g., `get_batch` -> `get_data`.
4. Add `torch.utils.data.get_worker_info` which returns worker information in a worker proc (e.g., worker id, dataset obj copy, etc.) and can be used in `IterableDataset.__iter__` and `worker_init_fn` to do per-worker configuration.
5. Add `ChainDataset`, which is the analog of `ConcatDataset` for `IterableDataset`.
6. Add `convert_fn`, which is meant to convert each loaded data into tensors. For the `MapWithBatchedRead` mode, fetched data is first converted using `convert_fn` and then collated. This shouldn't be much slower (if any) than the old approach of only using `collate_fn`.

### CODE REVIEW

File: torch/utils/data/dataset.py [hunk 1]
```diff
@@ -23,8 +24,26 @@ def __add__(self, other):
         return ConcatDataset([self, other])
 
 
+class IterableDataset(Dataset):
+    r"""An iterable Dataset.
+
+    All other datasets should subclass it. All subclasses should override
+    ``__len__``, that provides the size of the dataset, and ``__getitem__``,
```
Comment (kittipatv) [inline torch/utils/data/dataset.py]: > All subclasses should override ``__len__``

Is this a hard requirement? What if we have a stream of data?

Comment (SsnL) [reply torch/utils/data/dataset.py]: No it's not. I just copied from Dataset and haven't changed the doc. :) As long as you overwrite `__iter__`, you are good.

**Comment by apaszke** [comment]
Can you please elaborate on the final design of iterable datasets? As I've mentioned in some PRs to C++ Dataset API, I think that our current DataLoader story is completely mismatched with the desire to have mutable datasets (including iterable datasets). In particular, you can't simply replicate the dataset, and have all workers produce the same samples, because that would be silly. Similarly, the Sampler is completely useless in this case. I really think we should work on alternative solutions for this API.

File: test/test_dataloader.py [hunk 2]
```diff
@@ -231,30 +231,50 @@ def __len__(self):
         return self.size
 
 
+class WorkerSpecificIterableDataset(IterableDataset):
+    def __init__(self, sizes_for_all_workers):
+        self.sizes_for_all_workers = sizes_for_all_workers
+
+    def __iter__(self):
```
Comment (SsnL) [inline test/test_dataloader.py]: @apaszke Basically you can config each dataset replica differently in two different ways using `torch.utils.data.get_worker_info()` which returns the worker's `id`, `seed` and `dataset` replica:
1. In `__iter__` for an iterable dataset. 
2. In `worker_init_fn`.

This requires user to write their dataset code with multiprocessing data loading in mind. But I think this is a reasonable requirement because (1) there is no general way to split an iterator across multiple processes (2) if people want to do sharding / bulk loading, how to split work among workers should be part of the design.

**Review by apaszke — COMMENTED** [review]
Haven't read the new multiprocess iterator, but I had a few questions

File: torch/utils/data/_utils/worker.py [hunk 3]
```diff
@@ -55,7 +56,46 @@ def is_alive(self):
             return not self.manager_dead
 
 
-def _worker_loop(dataset, index_queue, data_queue, done_event, collate_fn, seed, init_fn, worker_id):
+WorkerInfo = namedtuple('WorkerInfo', ['id', 'seed', 'dataset'])
```
[the remark this answers is not in the document]
Comment (SsnL) [reply torch/utils/data/_utils/worker.py]: Yep I agree. Will make the change.

File: torch/utils/data/_utils/collate.py [hunk 4]
```diff
@@ -9,60 +9,89 @@
 import re
 from torch._six import container_abcs, string_classes, int_classes
 
+np_str_obj_array_pattern = re.compile(r'[SaUO]')
```
[the remark this answers is not in the document]
Comment (SsnL) [reply torch/utils/data/_utils/collate.py]: lol before #14668  we were using `re.search`, which is worse.

File: torch/utils/data/_utils/collate.py [hunk 5]
```diff
@@ -9,60 +9,89 @@
...
+            out_shape = torch.Size([len(batch)] + list(elem.size()))
+            if out is None:
+                out = elem.new_empty(out_shape)
+            else:
+                strides = [1]
+                for s in out_shape[1::-1]:
+                    strides.insert(0, strides[0] * s)
+                out = out.set_(out.storage, out_shape, strides)
+            for src, dst in zip(batch, out):
+                dst.copy_(src)
```
[the remark this answers is not in the document]
Comment (SsnL) [reply torch/utils/data/_utils/collate.py]: sure

File: torch/utils/data/dataloader.py [hunk 6]
```diff
@@ -70,47 +79,61 @@ class DataLoader(object):
     __initialized = False
 
     def __init__(self, dataset, batch_size=1, shuffle=False, sampler=None,
-                 batch_sampler=None, num_workers=0, collate_fn=default_collate,
+                 batch_sampler=None, num_workers=0,
+                 convert_fn=_utils.collate.default_convert,
```
[the remark this answers is not in the document]
Comment (SsnL) [reply torch/utils/data/dataloader.py]: With `collate_fn`, the input is assumed to be a list of **many data samples**, and you want to collate each field in the data structure of each element. E.g., `[([1, 2], [3, 4]), ([5, 6], [7, 8)]` should be come `(tensor([[1, 2], [5, 6]]), tensor([[3, 4], [7, 8]]))`. It does both elementwise conversion and collation.

However, when the dataset is an iterable, we want to only do conversion by considering the entire input as a single data sample. E.g., `[([1, 2], [3, 4]), ([5, 6], [7, 8)]` should be come a single `tensor([[[1, 2], [5, 6]], [[3, 4], [7, 8]]])`.

File: torch/utils/data/_utils/pin_memory.py [hunk 7]
```diff
@@ -35,23 +35,23 @@ def _pin_memory_loop(in_queue, out_queue, device_id, done_event):
...
+        return data.pin_memory()
+    elif isinstance(data, string_classes):
+        return data
+    elif isinstance(data, container_abcs.Mapping):
+        return {k: pin_memory_data(sample) for k, sample in data.items()}
+    elif isinstance(data, container_abcs.Sequence):
+        return [pin_memory_data(sample) for sample in data]
     else:
-        return batch
+        return data
```
[the remark this answers is not in the document]
Comment (SsnL) [reply torch/utils/data/_utils/pin_memory.py]: Because data loader can be used to load individual samples as well, and this PR provides better support for that use case. `batch` seems to assume that the data is always batched.

File: torch/utils/data/_utils/signal_handling.py [hunk 8]
```diff
@@ -33,8 +33,9 @@
 import signal
 import threading
 import torch
-from torch._C import _set_worker_pids, _remove_worker_pids, \
-    _error_if_any_worker_fails, _set_worker_signal_handlers
+from torch._C import _set_worker_pids, _remove_worker_pid, \
+    _remove_all_worker_pids, _error_if_any_worker_fails, \
+    _set_worker_signal_handlers
```
[the remark this answers is not in the document]
Comment (SsnL) [reply torch/utils/data/_utils/signal_handling.py]: nah, they are used as `signal_handling. _remove_all_worker_pids` in `dataloader.py`. I will add a comment.

File: torch/utils/data/_utils/worker.py [hunk 9]
```diff
@@ -75,9 +115,31 @@ def _worker_loop(dataset, index_queue, data_queue, done_event, collate_fn, seed,
 
         data_queue.cancel_join_thread()
 
+        global _worker_info
+        _worker_info = WorkerInfo(worker_id, seed, dataset)
+
         if init_fn is not None:
             init_fn(worker_id)
 
+        from torch.utils.data import DataLoaderMode
```
[the remark this answers is not in the document]
Comment (SsnL) [reply torch/utils/data/_utils/worker.py]: Sure :) We can always publicize it later if there's reasonable requirement.

File: torch/utils/data/_utils/worker.py [hunk 10]
```diff
@@ -87,23 +149,34 @@ def _worker_loop(dataset, index_queue, data_queue, done_event, collate_fn, seed,
...
-            idx, batch_indices = r
+            idx, index = r
             try:
-                samples = collate_fn([dataset[i] for i in batch_indices])
+                if mode == DataLoaderMode.Iterable:
+                    try:
+                        data = convert_fn(next(dataset_iter))
+                    except StopIteration:
+                        data = IterableDatasetStopIteration(worker_id)
+                        iteration_end = True  # don't waste resource in future iter
```
[the remark this answers is not in the document]
Comment (SsnL) [reply torch/utils/data/_utils/worker.py]: `try..except..`. The other important use of this flag is to prevent sending more than one `IterableDatasetStopIteration` to main proc. I will add this in the comment.

File: torch/utils/data/_utils/worker.py [hunk 11]
```diff
@@ -87,23 +149,34 @@ def _worker_loop(dataset, index_queue, data_queue, done_event, collate_fn, seed,
...
             except Exception:
                 # It is important that we don't store exc_info in a variable,
                 # see NOTE [ Python Traceback Reference Cycle Problem ]
                 data_queue.put((idx, ExceptionWrapper(sys.exc_info())))
             else:
-                data_queue.put((idx, samples))
-                del samples
+                data_queue.put((idx, data))
+                del data
+            del idx, index, r
```
[the remark this answers is not in the document]
Comment (SsnL) [reply torch/utils/data/_utils/worker.py]: will do :)

File: torch/utils/data/dataloader.py [hunk 12]
```diff
@@ -21,6 +24,12 @@
 default_collate = _utils.collate.default_collate
 
 
+class DataLoaderMode(object):
+    Map = "map_dataset"
+    MapWithBatchedRead = "map_dataset_with_batched_read"
+    Iterable = "iterable_dataset"
```
[the remark this answers is not in the document]
Comment (SsnL) [reply torch/utils/data/dataloader.py]: Sure. I'll make them ints!

File: torch/utils/data/dataloader.py [hunk 13]
```diff
@@ -34,6 +43,10 @@ class DataLoader(object):
             at every epoch (default: ``False``).
         sampler (Sampler, optional): defines the strategy to draw samples from
             the dataset. If specified, ``shuffle`` must be False.
+        drop_last (bool, optional): set to ``True`` to drop the last incomplete batch,
```
[the remark this answers is not in the document]
Comment (SsnL) [reply torch/utils/data/dataloader.py]: I moved it because it is only used together with `batch_size`, `shuffle` and `sampler`. Grouping them together seemed to make sense to me, but maybe I should respect the arg order more.

File: torch/utils/data/dataloader.py [hunk 14]
```diff
@@ -121,13 +144,91 @@ def __setattr__(self, attr, val):
...
+            return None
+        else:
+            return next(self.index_sampler_iter)  # may raise StopIteration
+
+    def __next__(self):
+        raise NotImplementedError
+
+    def __len__(self):
+        if self.mode == DataLoaderMode.Iterable:
+            raise NotImplementedError
```
[the remark this answers is not in the document]
Comment (SsnL) [reply torch/utils/data/dataloader.py]: Sure.

File: torch/utils/data/dataloader.py [hunk 15]
```diff
@@ -70,47 +79,61 @@ class DataLoader(object):
...
-
-        if batch_sampler is None:
-            if sampler is None:
-                if shuffle:
-                    sampler = RandomSampler(dataset)
-                else:
-                    sampler = SequentialSampler(dataset)
-            batch_sampler = BatchSampler(sampler, batch_size, drop_last)
-
-        self.sampler = sampler
```
[the remark this answers is not in the document]
Comment (SsnL) [reply torch/utils/data/dataloader.py]: Good point. Will use old name.

File: torch/utils/data/dataloader.py [hunk 16]
```diff
@@ -121,13 +144,91 @@ def __setattr__(self, attr, val):
...
+
+    def __next__(self):
+        if self.mode == DataLoaderMode.Iterable:
+            data = self.convert_fn(next(self.dataset_iter))  # may raise StopIteration
+        else:
+            index = self._next_index()  # may raise StopIteration
+            if self.mode == DataLoaderMode.Map:
+                data = self.convert_fn(self.dataset[index])
+            else:
+                # mode == DataLoaderMode.MapWithBatchedRead:
```
[the remark this answers is not in the document]
Comment (SsnL) [reply torch/utils/data/dataloader.py]: Because previous there is no easy way to to unbatch loading from a map-like dataset. One has to use `batch_size=1` and provide a custom collate_fn which is just `lambda x: x[0]`. That feels counterintuitive to me. With all other changes in this PR, this was a simple addition, so I made it.

**Comment by SsnL** [comment]
@apaszke There is no new multiprocessing iterator design. I just refactored it into two classes, each handling single or multi proc loading. The logic is exactly the same.

File: torch/utils/data/dataloader.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (SsnL) [reply torch/utils/data/dataloader.py]: Another reason is to provide better support for the case where the sampler provides a list of indices, and the dataset does bulk loading.

File: torch/utils/data/dataset.py [hunk 17]
```diff
@@ -44,9 +138,9 @@ def __len__(self):
 
 
 class ConcatDataset(Dataset):
-    """
-    Dataset to concatenate multiple datasets.
-    Purpose: useful to assemble different existing datasets, possibly
+    r"""Dataset to concatenate multiple datasets.
+
+    This classuseful to assemble different existing datasets, possibly
```
Comment (RicCu) [inline torch/utils/data/dataset.py]: nit: typo
```suggestion
    This class is useful to assemble different existing datasets, possibly
```

File: torch/utils/data/dataset.py [hunk 18]
```diff
@@ -87,8 +183,36 @@ def cummulative_sizes(self):
         return self.cumulative_sizes
 
 
-class Subset(Dataset):
+class ChainDataset(IterableDataset):
+    r"""Dataset to chain multiple :class:`IterableDataset` s.
+
+    This classuseful to assemble different existing dataset streamss, possibly
```
Comment (RicCu) [inline torch/utils/data/dataset.py]: nit: typos
```suggestion
    This class is useful to assemble different existing dataset streams, possibly
```

**Comment by concretevitamin** [comment]
@SsnL please take a look at https://github.com/pytorch/pytorch/issues/17451 to see if this proposal can support that use case!  /cc @soumith

**Comment by SsnL** [comment]
@concretevitamin Yes this would support that use case.

**Comment by concretevitamin** [comment]
@SsnL great, looking forward to seeing this PR getting merged!

File: docs/source/data.rst [hunk 19]
```diff
@@ -2,11 +2,219 @@ torch.utils.data
...
+The behavior of this mode is roughly equivalent with::
+
+    for index in sampler:
+        yield convert_fn(dataset[index])
+
+
+Loading from an iterable-style dataset
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+
```
Comment (faroit) [inline docs/source/data.rst]: It would be nice to have an example here so that people not familiar with the `utils.data` internals can still discuss the API and usage scenarios.

Comment (SsnL) [reply docs/source/data.rst]: Thanks for the suggestion. I agree. This part isn't finished yet. :)

Comment (faroit) [reply docs/source/data.rst]: no problem, take the time, its an important feature ;-)

I just thought it would be nice to do these docs in an early stage so that "we users" could comment on the API

**Comment by SsnL** [comment]
close in favor of https://github.com/pytorch/pytorch/pull/19228
